### PR TITLE
Create new entities: institution, region, country

### PR DESCRIPTION
**Pull request by andricDu**
Reverts icgc-argo/program-service#156

I accidentally fat fingered the merge button in the wrong window. recreating original PR with my comments rewritten bellow:

Rejecting this PR as it is unfinished.

Currently if I pass a create program request with body:
```json
{
  "program": {
    "short_name": {
      "value": "FAKE-CA"
    },
    "description": {
      "value": "FAKE-CA"
    },
    "name": {
      "value": "FAKE-CA"
    },
    "membership_type": {
      "value": 0
    },
    "commitment_donors": {
      "value": 10
    },
    "submitted_donors": {
      "value": 10
    },
    "genomic_donors": {
      "value": 10
    },
    "website": {
      "value": "https://example.com"
    },
    "cancer_types": [
      "FAKE"
    ],
    "primary_sites": [
      "Hello"
    ],
    "institutions": [
      "Hello"
    ],
    "countries": [
      "FAKE"
    ],
    "regions": [
      "Hello"
    ]
  },
  "admins": [
    {
      "email": {
        "value": "foobar@example.com"
      },
      "first_name": {
        "value": "Hello"
      },
      "last_name": {
        "value": "Hello"
      },
      "role": {
        "value": 0
      }
    }
  ]
}
```
It will create the program but it will leave the Cancer and Country fields empty as they are invalid.

There needs to be validation on input, especially considering that these need to be linked to existing Countries and Cancer types.

### CODE REVIEW

**Comment by andricDu** [comment]
References: https://github.com/icgc-argo/program-service/issues/129

File: src/main/java/org/icgc/argo/program_service/services/ProgramService.java [hunk 1]
```diff
@@ -118,67 +144,55 @@ public ProgramEntity createProgram(@NonNull Program program)
 
     val cancers = cancerRepository.findAllByNameIn(program.getCancerTypesList());
     val primarySites = primarySiteRepository.findAllByNameIn(program.getPrimarySitesList());
+    val institutions = institutionRepository.findAllByNameIn(program.getInstitutionsList());
+    val regions = regionRepository.findAllByNameIn(program.getRegionsList());
+    val countries = countryRepository.findAllByNameIn(program.getCountriesList());
```
Comment (andricDu) [inline src/main/java/org/icgc/argo/program_service/services/ProgramService.java]: A good example is here, where there is nothing to handle the case of returning an empty list. 

This should make a defensive check on whether or not valid countries were added. 


Something as simple as `countries.size() == program.getCountriesList().size()` could tell you if you had invalid countries entered.

File: src/main/resources/flyway/sql/V8__new_table_country_institution_region.sql [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  region_id           UUID NOT NULL,
+
+  PRIMARY KEY(program_id, region_id),
+  FOREIGN KEY(program_id) REFERENCES program(id),
+  FOREIGN KEY(region_id)  REFERENCES region(id)
+);
+
+ALTER TABLE PROGRAM
+DROP COLUMN institutions,
+DROP COLUMN countries,
```
Comment (andricDu) [inline src/main/resources/flyway/sql/V8__new_table_country_institution_region.sql]: This drops any existing information, and does not migrate it over to the new tables.

**Comment by andricDu** [comment]
The order of operations for the database migration is backwords. We should be creating the new tables with the populated values, creating associations between them and programs based on existing values, and then and only then do we want to drop the existing columns.

**Comment by andricDu** [comment]
@joneubank I added guards and transcation support on creation of a program. Good for you take take a look.

File: src/main/resources/flyway/sql/V9__populate_institution_country_region.sql [hunk 3]
```diff
@@ -0,0 +1,266 @@
+--Populate data for Institution, Country, and Region --
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+
+INSERT INTO institution (id, name)
+VALUES
+  (uuid_generate_v4(), 'OICR'),
+  (uuid_generate_v4(), 'FBI');
+
+INSERT INTO country (id, name)
+VALUES
```
Comment (joneubank) [inline src/main/resources/flyway/sql/V9__populate_institution_country_region.sql]: Are we not also validating the 2 character country codes in short names? Is this just done in the short_name validator and not matched from the DB?

File: src/main/java/org/icgc/argo/program_service/services/ProgramService.java [hunk 4]
```diff
@@ -18,207 +18,311 @@
...
+
+    if (cancers.size() != program.getCancerTypesList().size()
+      || primarySites.size() != program.getPrimarySitesList().size()
+      || institutions.size() != program.getInstitutionsList().size()
+      || countries.size() != program.getCountriesList().size()
+      || regions.size() != program.getRegionsList().size()) {
+      throw Status.INVALID_ARGUMENT
+        .augmentDescription(
+          "Cannot create program, Must provide valid cancer, primary site, institution, country, and region.")
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/services/ProgramService.java]: It is possible to write in a new institution name, I believe. Doesn't have to exist, can be created by this method.

File: src/main/java/org/icgc/argo/program_service/services/ProgramService.java [hunk 5]
```diff
@@ -18,207 +18,311 @@
...
+            .map(ps -> createProgramPrimarySite(programToUpdate, ps))
             .map(Optional::get)
             .collect(toUnmodifiableList());
     programPrimarySiteRepository.saveAll(programPrimarySites);
   }
 
+  private void processInstitutions(
+      @NonNull ProgramEntity programToUpdate, @NonNull List<String> names) {
+    val institutionEntities =
+        checkExistenceByName(InstitutionEntity.class, institutionRepository, names);
```
Comment (joneubank) [inline src/main/java/org/icgc/argo/program_service/services/ProgramService.java]: Same comment about institutions, we allow new names to be entered afaik.

File: src/test/java/org/icgc/argo/program_service/services/ProgramServiceTest.java [hunk 6]
```diff
@@ -24,79 +24,80 @@
...
-    val inputProgramEntity = new ProgramEntity().setName(RandomString.make(10)).setShortName(RandomString.make(33));
-    assertThat(inputProgramEntity.getCreatedAt()).isNull();
-    assertThat(inputProgramEntity.getUpdatedAt()).isNull();
-    when(programConverter.programToProgramEntity(program)).thenReturn(inputProgramEntity);
-    val outputEntity = programService.createProgram(program);
-    assertThat(outputEntity.getCreatedAt()).isNotNull();
-    assertThat(outputEntity.getUpdatedAt()).isNotNull();
-    verify(programRepository).save(inputProgramEntity);
-  }
+//  @Test
```
Comment (joneubank) [inline src/test/java/org/icgc/argo/program_service/services/ProgramServiceTest.java]: Remove commented code?

**Pull request merged**